File: Workshop/Day_6/S.java
import java.util.Scanner;

public class S
{
    public static int mp(int[] p, int f) {
        int n = p.length;
        if (n == 0) return 0;

        int h = -p[0];  
        int nh = 0;       

        for (int i = 1; i < n; i++) 
        {
            nh = Math.max(nh, h + p[i] - f);
            h = Math.max(h, nh - p[i]);
        }

        return nh;
    }

    public static void main(String[] args) 
    {
        Scanner scanner = new Scanner(System.in);

        System.out.print("No. of days: ");
        int n = scanner.nextInt();
        int[] p = new int[n];

        System.out.print("Stock Prices: ");
        for (int i = 0; i < n; i++) p[i] = scanner.nextInt();
        
        System.out.print("Transaction Fee: ");
        int f = scanner.nextInt();

        int ans = mp(p, f);
        System.out.println("The profit is " + ans);

        scanner.close();
    }
}

File: Workshop/Day_6/S.c
#include <stdio.h>

// Function to calculate maximum profit with a transaction fee
int calculateProfit(int* prices, int n, int fee) {
    int hold = -prices[0];  // Profit when holding a stock
    int noHold = 0;         // Profit when not holding a stock

    for (int i = 1; i < n; i++) {
        noHold = noHold > (hold + prices[i] - fee) ? noHold : (hold + prices[i] - fee);
        hold = hold > (noHold - prices[i]) ? hold : (noHold - prices[i]);
    }

    return noHold; // Maximum profit when not holding any stock
}

int main() {
    int n, fee;

    // Input number of days
    printf("Enter the number of days: ");
    scanf("%d", &n);

    if (n <= 0) {
        printf("Invalid number of days.\n");
        return 0;
    }

    // Input stock prices
    int prices[n];
    printf("Enter stock prices: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &prices[i]);
    }

    // Input transaction fee
    printf("Enter transaction fee: ");
    scanf("%d", &fee);

    // Calculate and print the maximum profit
    int profit = calculateProfit(prices, n, fee);
    printf("Maximum profit: %d\n", profit);

    return 0;
}
